optimade/server/entry_collections/entry_collections.py

```diff
@@ -153,10 +160,20 @@ def find(
                     detail=f"Instead of a single entry, {data_returned} entries were found",
                 )
 
-        exclude_fields = self.all_fields - response_fields
+        return results, data_returned, more_data_available, exclude_fields
+
+    def check_and_add_missing_fields(self, results: List[dict], response_fields: set):
+        """Checks whether the response_fields and mandatory fields are present.
+        If they are not present the values are set to None, so the deserialization works correctly.
+        It also checks whether all fields in the response have been defined either in the model or in the config file.
+        If not it raises an appropriate error or warning."""
         include_fields = (
             response_fields - self.resource_mapper.TOP_LEVEL_NON_ATTRIBUTES_FIELDS
-        )
+        ) | set(self.get_non_optional_fields())
+        # Include missing fields
+        for result in results:
+            for field in include_fields:
+                set_field_to_none_if_missing_in_dict(result["attributes"], field)
```

Comment:
I do not think this is particularly heavy compared to all the other things we do in the code.
It is what we previously did in handle_response_fields. I only moved it here, so we can do it before the deserialization and added code to handle nested fields.

For biomolecular data, a structure can easily have 10,000 atoms, so retrieving them from the database and putting them in the model would take some time. This way we can avoid that if the species_at_sites and cartesian_site_positions are not in the response_fields. (I also made a patch in the code for Barcelona that allowed them to specify the default response fields, so they can choose to not have these fields in the response by default.)

I did not want to make the change even bigger by bypassing the rest of the validator (as in your second suggestion). 
But from a performance viewpoint, bypassing the validation would be good.
Do you want me to add this to this PR or to a future PR ?  

I tried the root validator idea, but it seems I already get an error before the root_validator is executed, so I do not think this solution will work.

Reply:
Hmmm, fair enough, just looks a bit scarier as a double for loop alongside the recursive descent into dictionaries to get the nested aliases. It's quite hard to reason about this, so I might set up a separate repo for measuring performance in the extreme limits (1 structure of 10000 atoms vs 10000 structures of  a few atoms -- i.e., what we have now, ignoring pagination of course).

> I tried the root validator idea, but it seems I already get an error before the root_validator is executed, so I do not think this solution will work.

Did you use `@root_validator(pre=True)` to make sure it gets run before anything else? Perhaps bypassing the validation altogether can wait for another PR, as you suggest, but I'd like to make the performance tests first at least (doesn't necessarily hold up this PR but it might hold up the next release).

Reply:
I just noticed I made a mistake in my test script.
So it may be possible to do this with a root_validator after all.

Reply:
I have added a root_validator to the attributes class that checks whether all required fields are present and if not adds them and sets them to 0. I'll try to put the handling of the other include fields back to the place where it happened originally so the code changes less.